arch/arm/boot/dts/armada-385-ts7840.dts

```diff
@@ -0,0 +1,840 @@
+		};
+
+		/* XBEE */
+		fpga_uart8: fpga-uart8 {
+			compatible = "ns16550a";
+			reg = <0 0x140 8>;
+			interrupt-parent = <&fpgairqc>;
+			interrupts = <8 IRQ_TYPE_LEVEL_HIGH>;
+			clock-frequency = <1843200>;
+			reg-io-width = <1>;
+			fifo-size = <64>;
+		};
+
+		/* Spare */
+		fpga_uart9: fpga-uart9 {
+			compatible = "ns16550a";
+			reg = <0 0x148 8>;
+			interrupt-parent = <&fpgairqc>;
+			interrupts = <19 IRQ_TYPE_LEVEL_HIGH>;
+			clock-frequency = <1843200>;
+			reg-io-width = <1>;
+			fifo-size = <64>;
+		};
+
+		/* Spare */
+		fpga_uart10: fpga-uart10 {
+			compatible = "ns16550a";
+			reg = <0 0x150 8>;
+			interrupt-parent = <&fpgairqc>;
+			interrupts = <20 IRQ_TYPE_LEVEL_HIGH>;
+			clock-frequency = <1843200>;
+			reg-io-width = <1>;
+			fifo-size = <64>;
+		};
+
+		can@200 {
+			compatible = "nxp,sja1000";
+			reg = <0 0x200 0x00000100>;
+			interrupt-parent = <&fpgairqc>;
+			interrupts = <10 IRQ_TYPE_LEVEL_HIGH>;
+			reg-io-width = <1>;
+			nxp,external-clock-frequency = <16000000>;
+		};
+
+		opencores_spi0: spi@400 {
+			compatible = "opencores,spi-oc";
+			reg = <0 0x400 32>;
+			interrupt-parent = <&fpgairqc>;
+			interrupts = <11 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&fpga_pll0c1>;
+			clock-names ="spi-oc-clk";
+			opencores-spi,idx = <0>;
+			opencores-spi,num-chipselects = <1>;
+
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			spifram: spi@0 {
+				compatible = "cypress,fm25l16b", "atmel,at25";
+				reg = <0>;
+				spi-max-frequency = <19800000>;
+				size = <0x800>;
+				address-width = <16>;
+				pagesize = <64>;
+			};
+		};
+
+		opencores_spi1: spi@420 {
+			compatible = "opencores,spi-oc";
+			reg = <0 0x420 32>;
+			interrupt-parent = <&fpgairqc>;
+			interrupts = <12 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&fpga_pll0c1>;
+			clock-names ="spi-oc-clk";
+			opencores-spi,idx = <0>;
+			opencores-spi,num-chipselects = <1>;
+
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			spidev@0 {
+				compatible = "spidev";
+				reg = <0>;
+				spi-max-frequency = <1000000>;
+			};
+		};
+
+		mikro_i2c: i2c@500 {
+			compatible = "opencores,i2c-ocores";
+			reg = <0 0x500 4>;
+			interrupt-parent = <&fpgairqc>;
+			interrupts = <13 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&fpga_pll0c1>;
+			clock-names = "i2c-oc-clk";
+		};
+
+		mikro_pwm: pwm@600 {
+			compatible = "technologic,pwm";
+			reg = <0 0x600 8>;
+			clocks = <&fpga_pll0c1>;
+			clock-names = "pwm-input-clk";
+		};
+
+		ts_mmc: ts_mmc@800 {
+			compatible = "ts,tssdcore2";
+			reg = <0 0x800 32>;
+			//interrupt-parent = <&fpgairqc>;
+			//interrupts = <14 IRQ_TYPE_LEVEL_HIGH>;
+			busy-gpios = <&fpga_bank0 15 0>;
+			debug-gpios = <&fpga_bank1 12 0>;
+			status = "disabled";
+		};
+
+		mikro_adc0: adc@900 {
+			compatible = "technologic,ts-simple-adc";
+			reg = <0 0x900 4>;
+			interrupt-parent = <&fpgairqc>;
+			interrupts = <21 IRQ_TYPE_LEVEL_HIGH>;
+		};
+	};
+
+	sfp0: sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&sfp0_i2c>;
+	};
+
+	sfp1: sfp {
```

Comment:
These may need unique node names, e.g. `sfp0: sfp@0` since the node names are what the kernel uses to ID and the labels are intended more for human use